bitwarden: build from source

amarshall opened the pull request:
###### Description of changes

Bitwarden package is now built from source.

I’ve been running the source build for 11+ days now without any problems for my usage. In particular, though, testing of the functionality of the “desktop_native” feature would be appreciated, as that seems most likely to have gone awry—afaik I don’t use anything it does and it’s not clear to me how to exercise it manually.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x]  x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

amarshall committed:
bitwarden: build from source

NickCao commented on pkgs/tools/security/bitwarden/default.nix:
```diff
@@ -1,73 +1,117 @@
+    ];
+
+    # Requires /etc/machine-id
+    doCheck = false;
```
This can be hacked with libredirect: see https://github.com/NixOS/nixpkgs/pull/212682/files for example.

linyinfeng commented:
`bitwarden-bin` works well but `bitwarden` failed to build.

![image](https://user-images.githubusercontent.com/11229748/215736717-b8283dc0-3df2-42b8-b97a-088b19079819.png)

```console
$ nix run github:nixos/nixpkgs/pull/212712/head#bitwarden
error: builder for '/nix/store/s8niinrdv5q85fhsbhrvfdwqzsgkn92v-bitwarden-desktop-2023.1.1-npm-deps.drv' failed with exit code 127;
       last 7 log lines:
       > unpacking sources
       > unpacking source archive /nix/store/bkbnk2y7jzzixgqgyn2p3y7mpllcjkvh-source
       > source root is source
       > patching sources
       > /nix/store/vcin1kwfmn4izsm8lcn6706qygf6cqdk-stdenv-linux/setup: line 112: jq: command not found
       > /nix/store/vcin1kwfmn4izsm8lcn6706qygf6cqdk-stdenv-linux/setup: line 112: sponge: command not found
       > /nix/store/vcin1kwfmn4izsm8lcn6706qygf6cqdk-stdenv-linux/setup: line 129: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/s8niinrdv5q85fhsbhrvfdwqzsgkn92v-bitwarden-desktop-2023.1.1-npm-deps.drv'.
error: 1 dependencies of derivation '/nix/store/hsprymna30djpjnr6qy7f5lcwb9p1c72-bitwarden-desktop-2023.1.1.drv' failed to build
```

```console
$ nix log /nix/store/s8niinrdv5q85fhsbhrvfdwqzsgkn92v-bitwarden-desktop-2023.1.1-npm-deps.drv
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/bkbnk2y7jzzixgqgyn2p3y7mpllcjkvh-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
/nix/store/vcin1kwfmn4izsm8lcn6706qygf6cqdk-stdenv-linux/setup: line 112: jq: command not found
/nix/store/vcin1kwfmn4izsm8lcn6706qygf6cqdk-stdenv-linux/setup: line 112: sponge: command not found
/nix/store/vcin1kwfmn4izsm8lcn6706qygf6cqdk-stdenv-linux/setup: line 129: pop_var_context: head of shell_variables not a function context
```

dotlambda commented on pkgs/tools/security/bitwarden/default.nix:
```diff
@@ -1,73 +1,126 @@
-    comment = "A secure and free password manager for all of your devices";
-    desktopName = "Bitwarden";
-    categories = [ "Utility" ];
+  desktop-native = rustPlatform.buildRustPackage rec {
```
Does someone know whether the CLI will also use the Rust code at some point? In that case it might make sense to put this in its own file.

dotlambda commented on pkgs/tools/security/bitwarden/default.nix:
```diff
@@ -1,73 +1,126 @@
+
+  meta = common.meta // (with lib; {
+    maintainers = with maintainers; [ amarshall kiwi ];
     platforms = [ "x86_64-linux" ];
```
Shouldn't it support more platforms? You could relax this to `platforms.unix` to at least see what ofborg has to say about it.

NickCao commented:
The binary build keeps resource files under opt, makes it look cleaner. Can we do the same with the source build?

NickCao commented on pkgs/tools/security/bitwarden/default.nix:
```diff
@@ -1,73 +1,126 @@
+      dbus-run-session \
+        --config-file=${dbus}/share/dbus-1/session.conf \
+        -- bash -c cargoCheckHook
+      runHook postCheck
```
`export -f` is a neat trick! I was once hindered by the same issue (but on xvfb-run).

NickCao reviewed (approved):
LGTM (waiting a while before merging to see if others have any suggestions)

dotlambda commented:
What's the reason to keep bitwarden-bin around?

dotlambda commented on pkgs/desktops/gnome/core/gnome-keyring/default.nix:
```diff
@@ -78,7 +79,7 @@ stdenv.mkDerivation rec {
 
   # Use wrapped gnome-keyring-daemon with cap_ipc_lock=ep
-  postFixup = ''
+  postFixup = lib.optional useWrappedDaemon ''
```
```suggestion
  postFixup = lib.optionalString useWrappedDaemon ''
```

NickCao commented on pkgs/tools/security/bitwarden/default.nix:
```diff
@@ -1,73 +1,139 @@
-    desktopName = "Bitwarden";
-    categories = [ "Utility" ];
+  patches = [
+    ./remove-flaky-tests.patch
```
You can skip specific tests with "--skip=some::test::name", without using a patch.

dotlambda commented on pkgs/tools/security/bitwarden/default.nix:
```diff
@@ -1,73 +1,139 @@
 
-  buildInputs = [ libsecret libxshmfence ] ++ atomEnv.packages;
+  makeCacheWritable = true;
+  npmBuildFlags = "--workspace apps/desktop";
```
```suggestion
  npmBuildFlags = [
    "--workspace" "apps/desktop"
  ];
```

dotlambda commented:
> Either way, I’m happy to drop it now if that’s preferred.

I'd prefer that. But if someone plans on using the -bin version going forward we can keep it.

amarshall committed:
bitwarden: 2023.1.1 -> 2023.2.0

NickCao commented:
@ofborg build bitwarden-bin

NickCao reviewed (approved):
One last thing: please split the changes to gnome-keyring to it's own commit, also add comments/commit message about the motivation to do so.

NickCao commented:
If there are none known issues with the source build, I see no benefit in keeping the binary one. (Normally we keep binary builds for big-parallel packages like chromium, and bitwarden might not qualify?)

amarshall committed:
gnome-keyring: Add arg to build without wrapper

Some derivations may want to use gnome-keyring during their build, e.g.
for tests, but the wrapper for gnome-keyring-daemon not does not inside
the sandbox, so add an option to build without it. This will, at
least, be used in the upcoming bitwarden source-based build.
